# Лабораторная работа №2: Начало предварительной обработки данных. Формирование описания задачи








**Содержание**<a id='toc0_'></a>    
- 1. [Раздел 1. Обработка входных данных](#toc1_)    
  - 1.1. [Загрузка исходных данных](#toc1_1_)    
  - 1.2. [Анализ типов данных и значений признаков](#toc1_2_)    
  - 1.3. [Формирование обучающей и тестирующей выборок](#toc1_3_)    

# 1. <a id='toc1_'></a>[Раздел 1. Обработка входных данных](#toc1_)

### Постановка задачи
На основе предоставленной информации задача заключается в прогнозировании коэффициента `Гармония Бессмертия` (целевой признак), который отражает состояние магического ядра портала и его стабильность. Этот коэффициент является непрерывной величиной (пример значения: 0.975), что указывает на задачу регрессии. Цель — предсказать изменения этого коэффициента на основе входных параметров работы портала, чтобы вовремя вмешаться и предотвратить разрушение артефакта.

### Описание признакового пространства
#### Входные признаки
Все параметры, описывающие работу портала, кроме целевого и уникальных идентификаторов. Они включают числовые измерения (скорости, потоки, температуры, давления) и категориальные (тип вектора мощи).

#### Целевой признак
`Гармония Бессмертия` — непрерывный коэффициент стабильности (диапазон: [0, 1], где 1 — идеальное состояние, <1 — угасание).
#### Таблица с описанием признаков
| **№** | **Признак**               | **Описание**                                                                 | **Особенности** |
|-------|---------------------------|-----------------------------------------------------------------------------|-------------|
| 1     | Вектор Мощи              | Позиция магического механизма, управляющего интенсивностью перехода через портал. | >0          |
| 2     | Скорость перехода через портал | Скорость, с которой портал перемещает объект через пространство.            | >0          |
| 3     | Приток Силы Потока       | Поток магической силы, текущий через ядро портала, обеспечивая его стабильную работу. | >0          |
| 4     | Ритм магического ядра    | Число оборотов магического ядра портала в минуту.                           | >0          |
| 5     | Поток Энергий            | Число оборотов генератора эфира, преобразующего внешнюю энергию в магическую. | >0          |
| 6     | Сила Левого Потока       | Мощность магического потока, текущего через левую сторону портала, поддерживая его баланс. | >0          |
| 7     | Сила Правого Потока      | Мощность магического потока, текущего через правую сторону портала, обеспечивая равномерное распределение энергии. | >0          |
| 8     | Пламя Стихий             | Температура магической энергии, исходящей из высокого магического источника портала, в градусах Цельсия. | >0          |
| 9     | Температура вдоха Истока | Температура воздуха, входящего в магический ускоритель портала, в градусах Цельсия. | >0          |
| 10    | Температура выдоха Истока | Температура воздуха, исходящего из магического ускорителя портала, в градусах Цельсия. | >0          |
| 11    | Приток давления Выдоха Истока | Давление магического потока на выходе из магического источника высокого давления. | >0          |
| 12    | Давление вдоха Истока    | Давление воздуха, входящего в магический ускоритель.                        | >0          |
| 13    | Давление выдоха Истока   | Давление воздуха, исходящего из магического ускорителя.                     | >0          |
| 14    | Древний Ветер            | Давление древней магической энергии, покидающей портал в виде выхлопного потока. | >0          |
| 15    | Печать Чародея           | Параметр управления магическим впрыском в сердце портала, выраженный в процентах. | >=0          |
| 16    | Эмульсия Истока          | Количество магического топлива, подпитывающего портал.                     | >0          |
| 17    | Дыхание Истока           | Коэффициент, отображающий степень угасания магического ускорителя.          | >0          |
| 18    | Гармония Бессмертия - целевая переменная | Коэффициент, указывающий на состояние магического ядра портала и его стабильность, требующий магического восстановления в случае снижения. | >0|
| 19    | Тип Вектора Мощи         | Метка типа позиции магического механизма.                                  | Категориальный признак|
| 20    | Номер пометки            | Уникальный номер записи о работе портала в блакноте.                        | Нужно исключить этот признак|


#### **Формулы прикладной области**
1. **Общая мощность потоков**:
   - Сумма `Силы Левого Потока` и `Силы Правого Потока`:
     $$   P_{\text{потоки}} = \text{Сила Левого Потока} + \text{Сила Правого Потока}   $$
   - Совокупная сила всех потоков:
     $$   P_{\text{все потоки}} = \text{Сила Левого Потока} + \text{Сила Правого Потока} + \text{Приток Силы Потока}   $$

2. **Общая сила ядра**:
   $$   P_{\text{ядро}} = \text{Ритм магического ядра} \times \text{Приток Силы Потока}   $$

3. **Общее давление на выходе**:
   $$   P_{\text{давление}} = \text{Приток давления Выдоха Истока} + \text{Давление выдоха Истока}   $$

4. **Магическая производительность**:
   $$   \text{Производительность} = \frac{\text{Скорость перехода через портал}}{\text{Эмульсия Истока}}   $$

5. **Эффективность ядра**:
   $$   \text{Эффективность ядра} = \frac{P_{\text{ядро}}}{\text{Эмульсия Истока}}   $$

6. **Магическая мощность**:
   $$   P_{\text{магическая}} = \text{Эмульсия Истока} \times P_{\text{давление}} \times |\text{Пламя Стихий} - \text{Температура вдоха Истока}|   $$

7. **Степень износа магических источников**:
   $$   \text{Износ} = \frac{\text{Дыхание Истока}}{\text{Гармония Бессмертия}}   $$

8. **Расхождение в стабильности магии**:
   $$   \text{Расхождение} = |\text{Дыхание Истока} - \text{Гармония Бессмертия}|   $$

9. **Баланс угасания**:
   $$   \text{Баланс угасания} = \frac{\text{Дыхание Истока} - \text{Гармония Бессмертия}}{\text{Скорость перехода через портал}}   $$

#### Анализ зависимостей и формул
Следующие формулы могут быть использованы для создания новых признаков в задаче предсказания `Гармонии Бессмертия`:
- **Общая мощность потоков** (1): Учитывает ключевые параметры потоков, которые могут коррелировать с целевым признаком.
- **Общая сила ядра** (2): Отражает производительность ядра, влияющую на стабильность портала.
- **Общее давление на выходе** (3): Может быть связано с устойчивостью работы портала.
- **Магическая производительность** (4): Показывает эффективность использования топлива, что может влиять на стабильность.
- **Эффективность ядра** (5): Характеризует производительность ядра относительно топлива.
- **Магическая мощность** (6): Комплексный показатель, учитывающий давление и температуру, потенциально связанный с целевым признаком.

Следующие формулы нельзя использовать напрямую, так как они включают целевой признак (`Гармония Бессмертия`) в свои вычисления, что приведет к утечке данных в задаче регрессии:
- **Степень износа магических источников** (7): Использует `Гармонию Бессмертия` в знаменателе, что делает её непригодной для создания признаков.
- **Расхождение в стабильности магии** (8): Прямо использует целевой признак, что может привести к утечке данных.
- **Баланс угасания** (9): Также включает `Гармонию Бессмертия`, что исключает её из feature engineering.




**Формирование окружения для работы**

Первоначально необходимо подготовить окружение для работы с данными и их моделирования. Чтобы не возникало конфликтов виртуальной среды Python, зафиксируем используемые в ходе исследования версии основных библиотек и фреймворков, которые будут использоваться далее. А также пропишем последовательность команд для формирования среды.

Первоначально установим необходимые библиотеки при помощи пакетного менеджера `pip` в виртуальное окружение.

- Версия Python: 3.12.3

Сами версии можно получить с помощью команды:
`pip list --format=freeze`


In [ ]:
#!pip install --upgrade --force-reinstall -r requirements.txt

Импортируем библиотеки с псевдонимами:
| Наименование библиотеки | Описание                                                                             |
| ----------------------- | ------------------------------------------------------------------------------------ |
| matplotlib              | Библиотека для визуализации данных                                                   |
| pandas                  | Библиотека для работы таблицами и плоскими данными                                   |
| numpy                   | Библиотека для работы с массивами и матрицами                                        |
| scikit-learn            | Библиотека для построения классических моделей машинного обучения и работы с данными |
| seaborn                 | Библиотека для визуализации данных                                                   |
| typing                  | Библиотека для типизации парамеров функций и переменных                              |
| warnings                | Библиотека для корректировки выводов предупреждений                                  |


In [ ]:
import pandas as pd
import numpy as np
import typing as t
from sklearn.model_selection import train_test_split

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
import warnings
warnings.filterwarnings('ignore')

Для воспроизводимости результатов исследования зафиксируем основание генератора случайных чисел, который используется для деления выборки на обучение и тест, а также при обучении моделей.

In [ ]:
seed = 42
np.random.seed(seed)

## 1.1. <a id='toc1_1_'></a>[Загрузка исходных данных](#toc1_1_1_)

Данные в формате CSV с разделителем `|`. Загружаем с помощью `pd.read_csv()`.

In [ ]:
df = pd.read_csv('portal_data.csv', sep='|', encoding='cp1251')
df

Вектор Мощи Скорость перехода через портал  Приток Силы Потока  \
0            1.138                        1.54332             289.964   
1            2.088                              -            6960.180   
2            3.144             4.6299600000000005            8379.229   
3            4.161                        6.17328           14724.395   
4            5.140                         7.7166           21636.432   
...            ...                            ...                 ...   
11929        5.140                         7.7166           21624.934   
11930        6.175                        9.25992           29763.213   
11931        7.148                       10.80324           39003.867   
11932        8.206                       12.34656           50992.579   
11933        9.300                       13.88988           72775.130   

       Ритм магического ядра  Поток Энергий  Сила Левого Потока  \
0                 141.318158       6677.380               7.584   
1                 144.111767       6828.469              28.204   
2                 145.220853       7111.811              60.358   
3                 162.050156       7792.630             113.774   
4                 201.513586       8494.777             175.306   
...                      ...            ...                 ...   
11929             201.516623       8470.013             175.239   
11930             241.561772       8800.352             245.954   
11931             280.444950       9120.889             332.389   
11932             323.315332       9300.274             438.024   
11933             372.844216       9742.950             644.880   

       Сила Правого Потока  Пламя Стихий  Температура вдоха Истока  \
0                    7.584       464.006                       736   
1                   28.204       635.401                       736   
2                   60.358       606.002                       736   
3                  113.774       661.471                       736   
4                  175.306       731.494                       736   
...                    ...           ...                       ...   
11929              175.239       681.658                       736   
11930              245.954       747.405                       736   
11931              332.389       796.457                       736   
11932              438.024       892.945                       736   
11933              644.880      1038.411                       736   

      Температура выдоха Истока  Приток давления Выдоха Истока  \
0                             -                          1.096   
1                       581.658                          1.331   
2                             -                          1.389   
3                             -                          1.658   
4                       645.642                          2.078   
...                         ...                            ...   
11929                    628.95                          2.087   
11930                         -                          2.512   
11931                   680.393                          2.982   
11932                         -                          3.594   
11933                         -                          4.531   

      Давление вдоха Истока  Давление выдоха Истока Древний Ветер  \
0             Не определено                   5.947             -   
1                  1.394603                   7.282         1.019   
2             Не определено                   7.574             -   
3             Не определено                   9.007             -   
4             Не определено                  11.197         1.026   
...                     ...                     ...           ...   
11929         Не определено                  10.990         1.027   
11930         Не определено                  13.109             -   
11931         Не определено                  15.420         1.036   
1

## 1.2. <a id='toc1_2_'></a>[Анализ типов данных и значений признаков](#toc1_2_)

Проверяем типы данных с помощью `info()` и число уникальных значений с помощью `nunique()`.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11934 entries, 0 to 11933
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Вектор Мощи                     11934 non-null  float64
 1   Скорость перехода через портал  11934 non-null  object 
 2   Приток Силы Потока              11934 non-null  float64
 3   Ритм магического ядра           11934 non-null  float64
 4   Поток Энергий                   11934 non-null  float64
 5   Сила Левого Потока              11934 non-null  float64
 6   Сила Правого Потока             11934 non-null  float64
 7   Пламя Стихий                    11934 non-null  float64
 8   Температура вдоха Истока        11934 non-null  int64  
 9   Температура выдоха Истока       11934 non-null  object 
 10  Приток давления Выдоха Истока   11934 non-null  float64
 11  Давление вдоха Истока           11934 non-null  object 
 12  Давление выдоха Истока          

In [ ]:
df.nunique()

Вектор Мощи                           9
Скорость перехода через портал       10
Приток Силы Потока                11430
Ритм магического ядра              3888
Поток Энергий                     11834
Сила Левого Потока                 4286
Сила Правого Потока                4286
Пламя Стихий                      11772
Температура вдоха Истока              1
Температура выдоха Истока          2369
Приток давления Выдоха Истока       524
Давление вдоха Истока                 2
Давление выдоха Истока             4209
Древний Ветер                        20
Печать Чародея                     8496
Эмульсия Истока                     696
Дыхание Истока                       51
Гармония Бессмертия                  26
Тип Вектора Мощи                      4
Номер пометки                     11934
dtype: int64

In [ ]:
df.describe()

Вектор Мощи  Приток Силы Потока  Ритм магического ядра  Поток Энергий  \
count  11934.000000        11934.000000           11934.000000   11934.000000   
mean       5.166667        27247.498685             223.711688    8200.947312   
std        2.626388        22148.613155              81.061874    1091.315507   
min        1.138000          253.547000             136.939406    6589.002000   
25%        3.144000         8375.883750             145.220958    7058.324000   
50%        5.140000        21630.659000             201.514947    8482.081500   
75%        7.148000        39001.426750             280.447777    9132.606000   
max        9.300000        72784.872000             372.879926    9797.103000   

       Сила Левого Потока  Сила Правого Потока  Пламя Стихий  \
count        11934.000000         11934.000000  11934.000000   
mean           227.335768           227.335768    735.495446   
std            200.495889           200.495889    173.680552   
min              5.304000             5.304000    442.364000   
25%             60.317000            60.317000    589.872750   
50%            175.268000           175.268000    706.038000   
75%            332.364750           332.364750    834.066250   
max            645.249000           645.249000   1115.797000   

       Температура вдоха Истока  Приток давления Выдоха Истока  \
count                   11934.0                   11934.000000   
mean                      736.0                       2.352963   
std                         0.0                       1.084770   
min                       736.0                       1.093000   
25%                       736.0                       1.389000   
50%                       736.0                       2.083000   
75%                       736.0                       2.981000   
max                       736.0                       4.560000   

       Давление выдоха Истока  Печать Чародея  Эмульсия Истока  \
count            11934.000000    11934.000000     11934.000000   
mean                12.297123       33.641261         0.662440   
std                  5.337448       25.841363         0.507132   
min                  5.828000        0.000000         0.068000   
25%                  7.447250       13.677500         0.246000   
50%                 11.092000       25.276500         0.496000   
75%                 15.658000       44.552500         0.882000   
max                 23.140000       92.556000         1.832000   

       Дыхание Истока  Гармония Бессмертия  Номер пометки  
count     11934.00000           11934.0000   11934.000000  
mean          0.97500               0.9875    5966.500000  
std           0.01472               0.0075    3445.193391  
min           0.95000               0.9750       0.000000  
25%           0.96200               0.9810    2983.250000  
50%           0.97500               0.9875    5966.500000  
75%           0.98800               0.9940    8949.750000  
max           1.00000               1.0000   11933.000000

Удаляем `Номер пометки` как неинформативный идентификатор. Преобразуем строковые значения в NaN и приводим числовые столбцы к float. Также удаляем бесполезные признаки из-за их постоянных значений: `Температура вдоха Истока` и `Давление вдоха Истока`.

Несмотря на то что у столбцов `Поток Энергий` и `Пламя Стихий` много уникальных значений, их можно оставить, так как являются float значениями и имеют границы разбросов ~6000-9999 и ~400-1200 соответственно. А вот признак `Приток Силы Потока` имеет огромный разброс значений от 200 до 70000+, поэтому в будущем необходимо повторно оценить пользу этого признака.

In [ ]:
df.drop(columns=['Номер пометки'], inplace=True)
df = df.replace(['-', 'Не определено'], np.nan)
numeric_cols = df.columns.drop(['Тип Вектора Мощи', 'Гармония Бессмертия'])
for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')
df = df.drop(columns=['Температура вдоха Истока', 'Давление вдоха Истока'])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11934 entries, 0 to 11933
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Вектор Мощи                     11934 non-null  float64
 1   Скорость перехода через портал  10731 non-null  float64
 2   Приток Силы Потока              11934 non-null  float64
 3   Ритм магического ядра           11934 non-null  float64
 4   Поток Энергий                   11934 non-null  float64
 5   Сила Левого Потока              11934 non-null  float64
 6   Сила Правого Потока             11934 non-null  float64
 7   Пламя Стихий                    11934 non-null  float64
 8   Температура выдоха Истока       2387 non-null   float64
 9   Приток давления Выдоха Истока   11934 non-null  float64
 10  Давление выдоха Истока          11934 non-null  float64
 11  Древний Ветер                   2387 non-null   float64
 12  Печать Чародея                  

Теперь можно заметить, что признаки `Температура выдоха Истока` и `Древний Ветер` имеют 80% пропусков. Также у `Температура выдоха Истока` почти все значения уникальны, а у `Древний Ветер` - мало уникальных значений. Поэтому они могут быть бесполезны.

In [ ]:
df = df.drop(columns=['Температура выдоха Истока', 'Древний Ветер'])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11934 entries, 0 to 11933
Data columns (total 15 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Вектор Мощи                     11934 non-null  float64
 1   Скорость перехода через портал  10731 non-null  float64
 2   Приток Силы Потока              11934 non-null  float64
 3   Ритм магического ядра           11934 non-null  float64
 4   Поток Энергий                   11934 non-null  float64
 5   Сила Левого Потока              11934 non-null  float64
 6   Сила Правого Потока             11934 non-null  float64
 7   Пламя Стихий                    11934 non-null  float64
 8   Приток давления Выдоха Истока   11934 non-null  float64
 9   Давление выдоха Истока          11934 non-null  float64
 10  Печать Чародея                  11934 non-null  float64
 11  Эмульсия Истока                 11934 non-null  float64
 12  Дыхание Истока                  

Теперь можно увидеть, что осталось относительно небольшое количество пропусков в столбце `Скорость перехода через портал`.

In [ ]:
df.head()

Вектор Мощи  Скорость перехода через портал  Приток Силы Потока  \
0        1.138                         1.54332             289.964   
1        2.088                             NaN            6960.180   
2        3.144                         4.62996            8379.229   
3        4.161                         6.17328           14724.395   
4        5.140                         7.71660           21636.432   

   Ритм магического ядра  Поток Энергий  Сила Левого Потока  \
0             141.318158       6677.380               7.584   
1             144.111767       6828.469              28.204   
2             145.220853       7111.811              60.358   
3             162.050156       7792.630             113.774   
4             201.513586       8494.777             175.306   

   Сила Правого Потока  Пламя Стихий  Приток давления Выдоха Истока  \
0                7.584       464.006                          1.096   
1               28.204       635.401                          1.331   
2               60.358       606.002                          1.389   
3              113.774       661.471                          1.658   
4              175.306       731.494                          2.078   

   Давление выдоха Истока  Печать Чародея  Эмульсия Истока  Дыхание Истока  \
0                   5.947           7.137            0.082            0.95   
1                   7.282          10.655            0.287            0.95   
2                   7.574          13.086            0.259            0.95   
3                   9.007          18.109            0.358            0.95   
4                  11.197          26.373            0.522            0.95   

   Гармония Бессмертия Тип Вектора Мощи  
0                0.975           Слабый  
1                0.975           Слабый  
2                0.975    Ниже среднего  
3                0.975    Ниже среднего  
4                0.975    Выше среднего

## 1.3. <a id='toc1_3_'></a>[Формирование обучающей и тестирующей выборок](#toc1_3_)

Разделение данных на обучающую (80%) и тестовую (20%) выборки выбрано для обеспечения достаточного объема данных для обучения модели, сохраняя при этом репрезентативную тестовую выборку для оценки качества. Учитывая общий объем датасета (11934 записей), 80% (около 9547 записей) предоставляют достаточно примеров для обучения, а 20% (около 2387 записей) позволяют надежно проверить обобщающую способность модели.

Необходимо выделить обучающий и тестовый датасеты именно на данном этапе, чтобы не допустить утечки данных.

In [ ]:
y_columns = ['Гармония Бессмертия']
train_df, test_df = train_test_split(df, shuffle=True, random_state=seed, test_size=0.2)

X_train, X_test = train_df.drop(columns=y_columns), test_df.drop(columns=y_columns)
y_train, y_test = train_df[y_columns], test_df[y_columns]

In [ ]:
# Сохраняем обучающую и тестовую выборки
X_train.to_csv('X_train_lr2.csv', index=False)
X_test.to_csv('X_test_lr2.csv', index=False)
y_train.to_csv('y_train_lr2.csv', index=False)
y_test.to_csv('y_test_lr2.csv', index=False)

print("Выборки успешно сохранены для использования в ЛР №3")

Выборки успешно сохранены для использования в ЛР №3
